src/main/java/org/opensearch/knn/memoryoptsearch/faiss/FaissIdMapIndex.java

```diff
@@ -0,0 +1,242 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.knn.memoryoptsearch.faiss;
+
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+import org.apache.lucene.index.ByteVectorValues;
+import org.apache.lucene.index.FloatVectorValues;
+import org.apache.lucene.index.VectorEncoding;
+import org.apache.lucene.store.IndexInput;
+import org.apache.lucene.util.Bits;
+
+import java.io.IOException;
+
+/**
+ * A FAISS index with an ID mapping that maps the internal vector ID to a logical ID, along with the actual vector index.
+ * It first delegates the vector search to its nested vector index, then transforms the vector ID into a logical index that is
+ * understandable by upstream components. This is particularly useful when not all Lucene documents are indexed with a vector field.
+ * For example, if 70% of the documents have a vector field and the remaining 30% do not, the FAISS vector index will still assign
+ * increasing and continuous vector IDs starting from 0.
+ * However, these IDs only cover the sparse 30% of Lucene documents, so an ID mapping is needed to convert the internal physical vector ID
+ * into the corresponding Lucene document ID.
+ * If the mapping is an identity mapping, where each `i` is mapped to itself, we omit storing it to save memory.
+ */
+public class FaissIdMapIndex extends FaissIndex {
+    public static final String IXMP = "IxMp";
+
+    @Getter
+    private FaissHNSWFlatIndex nestedIndex;
+    private long[] vectorIdToDocIdMapping;
+
+    public FaissIdMapIndex() {
+        super(IXMP);
+    }
+
+    /**
+     * Partially load id mapping and its nested index to which vector searching will be delegated.
+     *
+     * @param input An input stream for a FAISS HNSW graph file, allowing access to the neighbor list and vector locations.
+     * @throws IOException
+     */
+    @Override
+    protected void doLoad(IndexInput input) throws IOException {
+        FaissIdMapIndex faissIdMapIndex = new FaissIdMapIndex();
+        readCommonHeader(input, faissIdMapIndex);
```

Comment:
I am not sure but with the change, it should be
```suggestion
      readCommonHeader(input, this);
```

Reply:
will do the next rev.